Tempo: Represent OTLP Span Intrinsics correctly

joey-grafana opened the pull request:
**What is this feature?**

Adds a span intrinsics section and moves all intrinsic tags to that section.

**Why do we need this feature?**

To represent OTLP Span Intrinsics correctly.

**Who is this feature for?**

Tempo users.

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/59607

This also fixes a few issues not reported;
- span kind internal not in downloaded trace file
- status.code of unset is ignored in uploaded file but exists in original trace
- tagsToAttributes(span.tags) throws if span.tags is undefined and export breaks

**Special notes for your reviewer:**

![Screenshot 2023-06-13 at 09 08 15](https://github.com/grafana/grafana/assets/90795735/1600bc65-6832-401c-8470-e5c0015e3ec7)



Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

joey-grafana committed:
Span intrinsics

joey-grafana commented on public/app/plugins/datasource/tempo/resultTransformer.ts:
```diff
@@ -161,51 +161,57 @@ function resourceToProcess(resource: collectorTypes.opentelemetryProto.resource.
+    if (span.status.code !== undefined) {
+      intrinsics.push({
         key: 'otel.status_code',
         value: SpanStatusCode[span.status.code],
       });
       if (span.status.message) {
-        spanTags.push({ key: 'otel.status_description', value: span.status.message });
+        intrinsics.push({ key: 'otel.status_description', value: span.status.message });
```
So we call these `status.code` & `status.message` in the backend but when transforming here we rename to `otel.status_code` and `otel.status_description` - not fully happy with that approach

adrapereira reviewed (dismissed):
In order to represent the OTLP intrinsics correctly we have to make a few more changes to this PR:
1. `error = true` shouldn't be displayed in the intrinsics section because it's not a valid intrinsic attribute. It may be a part of other sections if the system has a tag called `error` but in our case I think we're adding it "manually" to denote an error, which is wrong under OTLP. 
2. `span.kind` is not a valid intrinsic. `kind` is an intrinsic though so we can show it instead
3. same for `status.code`. However `status` is a valid intrinsic
4. Please review all the comments of https://github.com/grafana/grafana/issues/59607 to make sure we're making all the necessary and accurate changes.

aocenas commented:
I don't think it makes sense to hide the intrinsics as another table section. I think this comment with the image makes more sense: https://github.com/grafana/grafana/issues/59607#issuecomment-1416429862. Basically, intrinsics are well-known parts of the span that are always there while attributes is a variable length list of key-value pairs. This should be apparent from the UI and needs a different representation imho.

joey-grafana committed:
Update intrinsics and add to span details

joey-grafana committed:
Remove intrinsics section

joey-grafana committed:
Update tests

joey-grafana committed:
Update status code text

joey-grafana committed:
Self review

joey-grafana commented:
Thanks for the reviews :) 

I believe we had discussed in stand up that we were first going to move the intrinsics to their own section and then modify/rename them accordingly after some more discussion. In any case, I've updated this PR and removed the intrinsics section and instead display the values in the span details along with service, duration etc. 

I've removed `error` & `span.kind` has also been updated. `status` on the other hand needs to be handled differently. We assume that `status` is either `ok`, `error` or `unset` but [OTEL specifies](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#set-status) (+ [source](https://github.com/open-telemetry/opentelemetry-proto/blob/d8729d40f629dba12694b44c4c32c1eab109b00a/opentelemetry/proto/trace/v1/trace.proto#L260)) that `status` contains a `message` and a `code` and we also add each of these [as intrinsics](https://github.com/grafana/grafana/blob/main/pkg/tsdb/tempo/trace_transform.go#L284) on the span. The [values are represented](https://opentelemetry.io/docs/specs/otel/common/mapping-to-non-otlp/) as `otel.status_description` & `otel.status_code` (+ [source](https://github.com/search?q=repo%3Aopen-telemetry%2Fopentelemetry-go+otel.status_description&type=code)). This distinction is especially important when downloading/uploading traces.

joey-grafana committed:
Merge branch 'main' into joey/span-intrinsics

aocenas commented on public/app/plugins/datasource/tempo/resultTransformer.ts:
```diff
@@ -282,7 +289,8 @@ export function transformFromOTLP(
             duration: (span.endTimeUnixNano! - span.startTimeUnixNano!) / 1000000,
-            tags: getSpanTags(span, librarySpan.instrumentationLibrary),
+            tags: getSpanTags(span),
+            intrinsics: getIntrinsics(span, librarySpan.instrumentationLibrary),
```
I feel like this is a part I am having a different view of what intrinsics are so want to align or that. What I thought intrinsic means is just intrinsic attributes of the trace or span, not a different set of tags. So for example here `duration` is also intrinsic attribute because is it part of the span type. So this to me right now feel a bit odd in the same sense as if `startTime` and `duration` were part of some `intrinsics` map with `duration` key and value instead of just part of the actual type. Then when presenting and we have to have a separate list of keys, that we have to pull out of here first and then display it feels weird but at the same time reason I suggested to put it here because I thought they are just part of the type.

joey-grafana replied:
When you describe `duration` & `startTime` like that they do sound to be something that could go in the `intrinsics` map, however `duration` & `startTime` always exist on the span whereas not every span has the library intrinsics. Feels a bit tidier to put them in their own map in this sense. I wanted to keep them out of the tags map so we didn't have to filter when adding the tags to the UI however we could just put all of these on the span type and check if they exist before rendering in span details where they are now (next to duration, child count etc)

So to be super clear we would add the following to the span type;
```
kind: enum,
status: {
   code: number,
   message: string
},
instrumentationLibrary: {
   name: string,
   version: string 
}
```
btw have you ever seen the [trace state](https://github.com/grafana/grafana/pull/69394/files#diff-0fae7cc1f8c2a5f78b973620a1894c4a8e848581cc65e3ac2fc603b6a37fb64fL213) in a span?

aocenas replied:
> duration & startTime always exist on the span whereas not every span has the library intrinsics.

Ah ok yeah this is probably what I wasn't clear about. But in what cases we don't have them? I would assume the library and lib-version may be only from OTEL spans but kind and status are sort of universal, aren't they?

Also it's not like we show anything as intrinsic right? Like we have a hardcoded list of things that go there so it's not dynamic like somebody could send a span with some other intrinsics right?

joey-grafana replied:
Yes, they are so if they are set on the span they will be rendered.

That's it yeah, we have a hardcoded list of intrinsics that may vary in length depending on which exist on the span.

joey-grafana committed:
Move previously intrinsic values to span

joey-grafana committed:
Merge branch 'main' into joey/span-intrinsics

aocenas commented on public/app/plugins/datasource/tempo/resultTransformer.ts:
```diff
@@ -161,51 +161,75 @@ function resourceToProcess(resource: collectorTypes.opentelemetryProto.resource.
+
+function getStatusCode(span: collectorTypes.opentelemetryProto.trace.v1.Span) {
+  let statusCode = undefined;
+  if (span.status) {
```
Do we need these accessors here? Seems like they could be replaced by just using `?` operator like here `span.status?.code`?

aocenas reviewed (approved):
Small comment about the accessor functions but otherwise look good.

joey-grafana committed:
Remove few methods
